Fix resizing for rotated blocks

miina opened the pull request:
Fixes #2252.

miina commented on assets/src/components/resizable-box/index.js:
```diff
@@ -34,13 +46,84 @@ export default ( props ) => {
+				}
+
+				/*if ( 0 && angle ) {
+					switch ( direction ) {
```
@swissspidy If you have time, would you mind taking a look at this logic to see if you see if something seems to be off or if you have a better idea, perhaps you have a "fresher" eye. I was trying different options and for some reason, the calcuations didn't work as expected, not sure what I was missing. Even brought out a pencil and paper and everything seemed to make sense on paper, ha! I left in the previous tries and just commented it out to check again tomorrow.

Anyway, at this moment the code above kind of works, not smoothly though.

miina commented on assets/src/stories-editor/blocks/amp-story-page/edit.css:
```diff
@@ -126,7 +126,7 @@ div[data-type="amp/amp-story-page"] .editor-inner-blocks .editor-block-list__lay
 .editor-block-list__layout div[data-type="amp/amp-story-page"][data-amp-selected="parent"] .editor-block-drop-zone {
 	height: 583px;
-	width: 438px;
+	width: 558px;
```
Note: this is for extending the dropzone to be able to move blocks over the right edge of a page, too.

swissspidy replied:
So this is basically the size of the page plus some padding? Can we use some CSS variables for that to make this more clear?

miina replied:
Done in 33a7d6d.

miina commented on assets/src/components/higher-order/with-amp-story-settings.js:
```diff
@@ -231,6 +231,7 @@ export default createHigherOrderComponent(
 							isSelected={ isSelected }
```
There's another issue here -- image block has its own resizable box, however, that does not consider rotation. Any thoughts here? Should we try to hide the original one with CSS and replace it with the custom one instead?

swissspidy replied:
We did the same with the block movers already. I think there's no way around doing this for the resizable box as well. Go for it!

miina replied:
Just made this change within 3e7c44f.

swissspidy commented on assets/src/stories-editor/helpers/index.js:
```diff
@@ -1143,6 +1143,63 @@ export const maybeInitializeAnimations = () => {
+ */
+export const getDelta = ( event, angle, lastSeenX, lastSeenY, direction ) => {
+	let deltaH = 'bottom' === direction ? Math.sqrt( Math.pow( event.clientX - lastSeenX, 2 ) + Math.pow( event.clientY - lastSeenY, 2 ) ) : 0;
+	let deltaW = 'right' === direction ? Math.sqrt( Math.pow( event.clientX - lastSeenX, 2 ) + Math.pow( event.clientY - lastSeenY, 2 ) ) : 0;
```
It might be useful to have small helper function for this, e.g. something like this:

```js
const getDelta = ( deltaX, deltaY ) => Math.sqrt( deltaX * deltaX + deltaY * deltaY );

let deltaH = 'bottom' === direction ? getDelta( event.clientX - lastSeenX, event.clientY - lastSeenY ) : 0;
let deltaW = 'right' === direction ? getDelta( event.clientX - lastSeenX, event.clientY - lastSeenY ) : 0;
```

It's not only DRY but also implicitly documented.

swissspidy commented:
A few observations:

* While resizing a rotated element it can be somewhat shaky, i.e. moving for a few sub-pixels  
  Happens with [this other React library](https://mockingbot.github.io/react-resizable-rotatable-draggable/) too, so I guess there's no way to avoid it
* Sometimes the resize handle jumps a bit, see video below for a demo
* Resizing an image crops it off instead of proportionally resizing it like in the regular posts editor
* It might make sense to have resize handlers on every side, so you can easily resize a block even when it's rotated in a weird way
* Some tests for these new helper functions would be nice, and they would also help with documentation

Video: https://cloudup.com/cgkCeM4qlHg

Editor view end result:

<img width="373" alt="Screenshot 2019-05-15 at 17 47 21" src="https://user-images.githubusercontent.com/841956/57818696-f17f3e00-7739-11e9-88e7-64dac20d9c9c.png">

Frontend view end result:

<img width="452" alt="Screenshot 2019-05-15 at 17 47 26" src="https://user-images.githubusercontent.com/841956/57818699-f512c500-7739-11e9-82e9-65e21e6ee401.png">

miina commented:
@swissspidy Thank you for testing!

> While resizing a rotated element it can be somewhat shaky, i.e. moving for a few sub-pixels
Happens with this other React library too, so I guess there's no way to avoid it

Yes, it seems like it's not possible to avoid it fully.

> Sometimes the resize handle jumps a bit, see video below for a demo

Not sure exactly why that happens, it seems to happen around the spot where the resizing event starts, looking into this.

> Resizing an image crops it off instead of proportionally resizing it like in the regular posts editor

This should be fixed now.

> It might make sense to have resize handlers on every side, so you can easily resize a block even when it's rotated in a weird way

Maybe that's something that we could add once the drag handle has been removed? I'll look into that as the last thing once the rest already works. It will need some additional calculations.

> Some tests for these new helper functions would be nice, and they would also help with documentation

Good point! Once it actually works well enough, will add the tests.

miina commented:
@swissspidy I've added some tests and hopefully addressed most of your feedback. Could you give it another try? Note that I didn't add the `top` and `left` resizing right now, this would need reworking the calculation for the new positioning of the block and I'm not sure if I'll have enough time in time. Also, in some cases the drag handle would collide with the left resizer so we could potentially implement it together with #2275.

swissspidy commented:
Awesome stuff! 🎉 

A few observations for this round of testing:

Lowering the height too much results in the block's position being changed in an odd way. Maybe we can enforce some limit here?

Video: https://cloudup.com/iyj1lYp7_Ow

Image still seems to get cropped. Video: https://cloudup.com/iPfn0sY6cii

Also, the resizing isn't as smooth as in the regular post editor. Video: https://cloudup.com/ichQUmFl3mW

Not sure if fixing that one is easily possible.

swissspidy commented:
Pinging @westonruter @kienstra for some testing as well.

miina commented:
> Image still seems to get cropped.

Actually, I think it should now be the same behavior as in the FE -- it's using `cover`. For example:
FE:
![Screen Shot 2019-05-17 at 12 02 39 AM](https://user-images.githubusercontent.com/3294597/57890199-40f76400-7837-11e9-81f1-5dab3eacc13b.png)

Editor:
![Screen Shot 2019-05-17 at 12 02 35 AM](https://user-images.githubusercontent.com/3294597/57890200-418ffa80-7837-11e9-9d4f-5ed9d3a58495.png)

Original image size:
![Screen Shot 2019-05-17 at 12 04 49 AM](https://user-images.githubusercontent.com/3294597/57890254-68e6c780-7837-11e9-95a3-a846bd971acb.png)

This should be the related PR: #2250.

Thoughts?

> Lowering the height too much results in the block's position being changed in an odd way. Maybe we can enforce some limit here?

Hmm, that shouldn't probably happen, will try to fix.

> Also, the resizing isn't as smooth as in the regular post editor. Video: https://cloudup.com/ichQUmFl3mW
Not sure if fixing that one is easily possible.

At this moment I'm not sure how to make it smoother. We could maybe open a separate bug issue for that if it seems to be distracting the editing flow. Thoughts?

swissspidy commented:
> At this moment I'm not sure how to make it smoother. We could maybe open a separate bug issue for that if it seems to be distracting the editing flow.

Works for me. Resizing also affects the `amp-fit-text` font size calculation, so I guess we can tackle these two things together.

swissspidy commented:
As for the image size, I guess it makes sense then.

miina commented on assets/src/components/resizable-box/index.js:
```diff
@@ -34,13 +47,64 @@ export default ( props ) => {
+					};
+
+					// @todo Figure out why calculating the new top / left position doesn't work in case of small height value.
+					// @todo Remove this temporary fix.
```
@swissspidy I changed the max and min sizes for resizing based on the values in your PR #2296. Also added this check here for when height is less than 60, since that's when the shift starts, however, I'm not sure why it happens. Added this fix here for now since might not have time to figure out the reason behind it. Let me know if it seems to work a bit better now.

westonruter commented:
I'm getting an error when I select an image from the media library:

> RangeError: Maximum call stack size exceeded

![image](https://user-images.githubusercontent.com/134745/57955051-966c5900-78a9-11e9-92f6-508ef4ab58c2.png)

Testing in Gutenberg 5.7.

miina commented:
@westonruter I'm getting the same error on the `amp-stories-redux` branch, this doesn't seem to be specific to this PR.

Created a separate issue for that: https://github.com/ampproject/amp-wp/issues/2350

westonruter commented:
@miina Thanks, the error is fixed.

Otherwise, I'm having a difficult time moving the block. It seems I have to deselect the block and re-select it in order to be able to drag the block with the handle. If I don't rotate or resize, then I can drag immediately after selecting. But if I resize or rotate the block, then the image bounces back to the original location after dragging (even though the ghost does move as I drag).

miina commented:
@westonruter This issue will get fixed within: https://github.com/ampproject/amp-wp/issues/2306
It's also a general and not this PR-specific issue.

westonruter commented:
@miina Got it. Thanks. One other thing: I noticed with the changes in this PR, an image's actual size does not change until I let go of the resize handle. Prior to this PR, the image would resize _while_ I move the resize handles. So this new behavior feels less user-friendly in how the image appears during resizing. Is that fixable?

miina commented:
@westonruter Good catch. We could probably create special handling for the image block to assign the current value to its wrapper while resizing, it should be possible.

We could create a separate bug issue for that, alternatively I could look into it tomorrow within this PR (won't have time tonight). Thoughts?
